external/avro/src/main/scala/org/apache/spark/sql/avro/AvroUtils.scala

```diff
@@ -228,6 +236,38 @@ private[sql] object AvroUtils extends Logging {
       .groupBy(_.name.toLowerCase(Locale.ROOT))
       .mapValues(_.toSeq) // toSeq needed for scala 2.13
 
+    private[this] val fieldPairs = catalystSchema
+      .zipWithIndex
+      .map(fieldWithPos => (fieldWithPos, getAvroField(fieldWithPos._1.name, fieldWithPos._2)))
+
+    /**
+     * Get the fields which have matching equivalents in both Avro and Catalyst schemas.
+     *
+     * @return A sequence of `(catalystField, catalystPosition, avroField)` tuples.
+     */
+    def getMatchedFields: Seq[(StructField, Int, Schema.Field)] = fieldPairs.flatMap {
+      case ((sqlField, sqlPos), Some(avroField)) => Some((sqlField, sqlPos, avroField))
+      case (_, None) => None
+    }
+
+    def assertNoExtraSqlFields(includeNullable: Boolean): Unit =
+      fieldPairs.filter(_._2.isEmpty).map(_._1).filter(includeNullable || !_._1.nullable) match {
```

Comment:
TBH, this line is a bit hard to understand...

Reply:
Agreed, I refactored this to make it more clear.